students/ArchWu/Session4/mailroom.py

```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env python3
+import sys
+import operator
+import tempfile
+#donations[]
+#donors {donor1:donations1, donor2:donations2}
+donors = {}
+# The data struction used in this program, a list of tuples
+
+
+def add_donor(name, donation = 0):
+    """ Adds a new donor with an initial donation """
+    donations = [donation]
+    donors[name] = donations
+    return
+
+def add_donation(name, donation):
+    """ Adds a donation to an existing donor """
+    donors[name].append(donation)
+    return
+
+def report():
+    """ Make a report on donations received by a style as specified """
+    print("Donor Name                | Total Given | Num Gifts | Average Gift")
+    print("------------------------------------------------------------------")
+    sorted_donors = sorted(donors.items(), reverse = True, key = lambda item: (sum(item[1])))
+    for donor in sorted_donors:
+        name, total, num_gifts, avg = gen_stats(donor)
+        print("{:25s}   {:11.2f}   {:9d}   {:12.2f}".format(name, total, num_gifts, avg))
+    return
+
+def gen_stats(donor):
+    """ A function calculates the total and average donation of a donor """
+    name = donor[0]
+    donations = donor[1]
+    total = round(sum(donations),2)
+    num_gifts = len(donations)
+    avg = round((total / num_gifts), 2)
+    return (name, total, num_gifts, avg)
+
+def thank_you():
+    """ Thank-you funcion that generates an email to thank new donations """
+    while True:
+        answer = input("Please type the Full Name, type quit to go back to previous menu> ")
+        if answer == 'list':
+            for donor in donors:
+                print(donor[0])
```

Comment:
Be careful here, we are no longer using a tuple so we are not printing the 0th index of the string. We want the entire key 😄